observer_pattern.java

```diff
@@ -0,0 +1,77 @@
+import java.util.ArrayList;
+import java.util.List;
+
+// Subject interface
+interface Subject {
+    void addObserver(Observer observer);
+    void removeObserver(Observer observer);
+    void notifyObservers();
+}
+
+// Concrete Subject
+class ConcreteSubject implements Subject {
+    private List<Observer> observers = new ArrayList<>();
+    private int state;
+
+    public int getState() {
+        return state;
+    }
+
+    
+    public void setState(int state) {
+        this.state = state;
+       
+    }
+
+    @Override
+    public void addObserver(Observer observer) {
+        observers.add(observer);
+    }
+
+    @Override
+    public void removeObserver(Observer observer) {
+        observers.remove(observer);
+    }
+
+    @Override
+    public void notifyObservers() {
+        for (Observer observer : observers) {
+            observer.update();
+        }
+    }
+}
+
+// Observer interface
+interface Observer {
+    void update();
+}
+
+// Concrete Observer
+class ConcreteObserver implements Observer {
+    private ConcreteSubject subject;
+
+    public ConcreteObserver(ConcreteSubject subject) {
+        this.subject = subject;
+        this.subject.addObserver(this);
+    }
+
+    @Override
+    public void update() {
+        System.out.println("Observer updated. New state: " + subject.getState());
+    }
+}
+
+public class Main {
+    public static void main(String[] args) {
+        ConcreteSubject subject = new ConcreteSubject();
+        ConcreteObserver observer1 = new ConcreteObserver(subject);
+        ConcreteObserver observer2 = new ConcreteObserver(subject);
```

Comment:
Subscription of observers to the subject should happen here, they should not occur inside the observer constructor. In other words, you should have something similar to the following here:

```java
subject.addObserver(observer1)
```